util/benchmark/ec/p256_awslc_ossl/src/benchmark_ecdh.c

```diff
@@ -0,0 +1,183 @@
+        if (!(test_ctx = EVP_PKEY_CTX_new(key_B, NULL)) || /* test ctx from skeyB */
+            !EVP_PKEY_derive_init(test_ctx) || /* init derivation test_ctx */
+            !EVP_PKEY_derive_set_peer(test_ctx, key_A) || /* set peer pubkey in test_ctx */
+            !EVP_PKEY_derive(test_ctx, NULL, &test_outlen) || /* determine max length */
+            !EVP_PKEY_derive(ctx, secret_a, &outlen) || /* compute a*B */
+            !EVP_PKEY_derive(test_ctx, secret_b, &test_outlen) || /* compute b*A */
+            test_outlen != outlen /* compare output length */ ||
+            CRYPTO_memcmp(secret_a, secret_b, outlen)) /* compare A's and B's shared secret bytes */
+        {
+            ecdh_checks = 0;
+            BIO_printf(bio_err, "ECDH shared secret test failure.\n");
+            ERR_print_errors(bio_err);
+        }
+    }
+
+    if (1 == ecdh_checks)
+    {
+        /* Test the key derivation after regenerating A's key using the same context */
+        if (EVP_PKEY_keygen(kctx, &key_A) <= 0 || /* generate secret key A again */
+            !EVP_PKEY_derive(ctx, secret_a, &outlen) || /* compute a*B */
+            !CRYPTO_memcmp(secret_a, secret_b, outlen) || /* ensure not same shared secret as before */
+            !EVP_PKEY_derive(test_ctx, secret_b, &test_outlen) || /* compute b*A */
+            CRYPTO_memcmp(secret_a, secret_b, outlen)) /* compare A's and B's shared secret bytes */
+        {
+            ecdh_checks = 0;
+            BIO_printf(bio_err, "ECDH shared secret with new A key test failure.\n");
+            ERR_print_errors(bio_err);
+        }
+    }
+
+    if (1 == ecdh_checks)
+    {
+        /* Warm up and instrument the function to calculate how many iterations it should run for */
+        start = time_now();
+        /* key generation and key derivation on A's side */
+        for (int i = 0; i < WARM_UP_NUM_ITER; i++)
+        {
+            EVP_PKEY_keygen(kctx, &key_A);
+            EVP_PKEY_derive(ctx, secret_a, &outlen);
+        }
+        end = time_now();
+        num_itr = calculate_iterations(start, end, WARM_UP_NUM_ITER, usec);
+
+        start = time_now();
+
+        /* Benchmark key generation and key derivation on A's side */
+        for (int i = 0; i < num_itr; i++)
+        {
+            EVP_PKEY_keygen(kctx, &key_A);
+            EVP_PKEY_derive(ctx, secret_a, &outlen);
+        }
+        end = time_now();
+
+        report_results(start, end, num_itr, "ECDH P-256");
+    }
+
+    if (1 == ecdh_checks)
+    {
+        BIO_printf(bio_out, "SUCCESS: ECDH Test.\n");
+    }
+
+    EVP_PKEY_free(key_A);
+    EVP_PKEY_free(key_B);
+    EVP_PKEY_CTX_free(kctx);
+    kctx = NULL;
+    EVP_PKEY_CTX_free(ctx);
+    ctx = NULL;
+    EVP_PKEY_CTX_free(test_ctx);
+    test_ctx = NULL;
+
+    if (NULL != secret_a)
+    {
+        OPENSSL_free(secret_a);
+    }
+
+    if (NULL != secret_b)
+    {
+        OPENSSL_free(secret_b);
+    }
+
+	return;
+}
```

Comment:
If I understood correctly, the first part of this function is testing the correctness of the implementation, and only after this test comes the actual benchmark. If that's the case, maybe a comment should be added just to note this.

Reply:
Would you suggest where a comment is missing, Dusan, and what it should say. When I reviewed the file, I found comments before the different parts. Thank you.

Reply:
What I meant is to add a comment in the beginning (line 34) that says "We first test the correctness of the implementation (initialization of parameters, key generation, ...). If all the tests pass we continue to the benchmarking part of the function." And then in line 131 just before the benchmark, to add a comment that benchmarking starts there.

Reply:
Done. Thank you for the suggestion.